### PR TITLE
CQL calculation logic

### PR DESCRIPTION
**Pull request by ljades**
Implements the CQL calculation logic needed to calculate on patients against measures.

**Submitter:**
- [x] This pull request describes why these changes were made.
- [x] Internal ticket for this PR: https://jira.mitre.org/browse/TACO-45, https://jira.mitre.org/browse/TACO-50
- [x] Internal ticket links to this PR
- [x] Code diff has been done and been reviewed
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass

**Reviewer 1:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

**Reviewer 2:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

### CODE REVIEW

**Review by holmesie — APPROVED** [review]

File: lib/executor.js [hunk 1]
```diff
@@ -1,57 +1,103 @@
...
+
+    handler.start(options);
+    await this.valueSetSource.findValueSetsByMongoidForMeasures(measures, (self) => {
+      executor.executeForEachMeasure(measures, self.valueSetsByMongoid, handler, options);
+    });
+    return handler.finish();
+  }
+
+  executeForEachMeasure(measures, valueSetsByMongoid, handler, options) {
     measures.forEach((mes) => {
```
Comment (jbradl11) [inline lib/executor.js]: I am just starting to wrap my head around how this repo is structured so I am not ready to add a bunch of comments. However, I believe "mes" should be changed to something more specific like "measure"

Comment (ljades) [reply lib/executor.js]: Originally I was debating between using "measure" or "mes". I think I ended up choosing "mes" because I felt it would be a more readable distinction from multiple "measures". Either is fine, just wanted to make aware of the original thought process.

(Also, hi! I'm not active on the project anymore but I'm still here to answer questions if any come up!)

File: lib/handlers/cypress_mongo_handler.js [hunk 2]
```diff
@@ -0,0 +1,168 @@
+const mongoose = require('mongoose');
+const IndividualResultSchema = require('cqm-models').IndividualResultSchema;
+
+module.exports = class Handler {
```
Comment (jbradl11) [inline lib/handlers/cypress_mongo_handler.js]: There should be some type of super class "Handler" which specific handlers "extend" from.  I recommend naming this to something more cypress specific.

File: lib/models/cql_calculator.js [hunk 3]
```diff
@@ -0,0 +1,643 @@
...
+  }
+
+  // Key for storing calculators on a measure
+  static calculationKey(measure) { return `${measure.id}`; }
+
+  // Key for storing results for a patient / population calculation
+  // we use the mongo ID for the patient portion
+  cacheKey(measure, patient) { return `${this.calculationKey(measure)}/${patient.id()}`; }
+
+  clearResult(measure, patient) {
```
Comment (jbradl11) [inline lib/models/cql_calculator.js]: I don't believe `calculationKey`, `cacheKey`, or `clearResult` are ever used. Unless there is a plan to use them in the future they should be removed and if there is a plan to use them in the future, add a comment explaining that.

File: lib/models/cql_calculator.js [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+    // We store both the calculation result and the calculation code based on keys derived from the arguments
+    let generatedELMJSON;
+
+    const resultsByPatient = {};
+
+    this.valueSetsByMongoid = valueSetsByMongoid;
+
+    patientSource.reset();
+
+    // try {
```
Comment (jbradl11) [inline lib/models/cql_calculator.js]: Remove, or add back in the try / catch block. There should probably be some type of error handling.

File: lib/models/cql_calculator.js [hunk 5]
```diff
@@ -0,0 +1,643 @@
...
+      } else {
+        end = this.constructor.getConvertedTimeEndOfYear(options.effective_date);
+      }
+    } else {
+      start = this.constructor.getConvertedTime(measure.get('measure_period').low.value);
+      end = this.constructor.getConvertedTime(measure.get('measure_period').high.value);
+    }
+    const startCql = Cql.DateTime.fromDate(start, 0); // No timezone offset for start
+    const endCql = Cql.DateTime.fromDate(end, 0); // No timezone offset for stop
+
```
Comment (jbradl11) [inline lib/models/cql_calculator.js]: There is another "parameter" that was added in the Bonnie calculator. `executionDateTime = cql.DateTime.fromDate(new Date(), '0')`  This creates the execution DateTime that we pass into the engine. Essentially telling the engine what the current time is when the execution starts. There is more code that will need to be updated further down in this file.

File: lib/models/cql_calculator.js [hunk 6]
```diff
@@ -0,0 +1,643 @@
...
+          // In episode of care based measures, episodeResults contains the population results
+          // for EACH episode.
+          result.set({ episode_results: episodeResults });
+        }
+        result.patient = patientId;
+        result.measure = measure._id;
+        resultsByPatient[patientId] = result;
+      }
+    });
+    // } catch (error) {
```
Comment (jbradl11) [inline lib/models/cql_calculator.js]: Same as my comment above.

File: lib/models/cql_calculator.js [hunk 7]
```diff
@@ -0,0 +1,643 @@
...
+    } else if ((populationResults.DENEX != null) &&
+               !this.constructor.isValueZero('DENEX', populationResults) &&
+               (populationResults.DENEX >= populationResults.DENOM)) {
+      if ('NUMER' in populationResults) {
+        populationResultsHandled.NUMER = 0;
+      }
+      if ('NUMEX' in populationResults) {
+        populationResultsHandled.NUMEX = 0;
+      }
+    // TODO: Temporarily disable the removal of OBSERVs when a member of MSRPOPLEX as to not change actual result values uncomment for 2.1 release
```
Comment (jbradl11) [inline lib/models/cql_calculator.js]: I believe this is safe to uncomment.

File: lib/models/cql_calculator.js [hunk 8]
```diff
@@ -0,0 +1,643 @@
...
+      }
+    });
+    // } catch (error) {
+    //   throw Error(error);
+    // }
+
+    return resultsByPatient;
+  }
+
+  static executeEngine(elm, patientSource, valueSets, libraryName, version, parameters = {}) {
```
Comment (jbradl11) [inline lib/models/cql_calculator.js]: Either pass in the new `executionDateTime` that I mentioned above into this function and pass it into `executor.exec(patientSource, executionDateTime)` or set `executionDateTime` in this function. This function could probably be refactored. It is currently a copy of a function that Bonnie calls in the CQL execution engine repo called `executeSimpleELM`, now that its being brought into this repo I think it would make sense to move around some of the logic. Thoughts?

File: lib/executor.js [hunk 9]
```diff
@@ -1,57 +1,103 @@
...
+  async execute(patientIds, measureIds, options = null) {
+    await this.getPatientsById(patientIds);
+    const measures = await this.getMeasuresById(measureIds);
+
+    const executor = this;
+
+    let handler;
+
+    if (!options || !options.handlerType || options.handlerType === 'CypressMongo') {
+      handler = new CypressHandler(this.connectionInfo);
```
Comment (jbradl11) [inline lib/executor.js]: Is there a way to make this more flexible to better handle different "HandlerTypes"?

File: lib/models/patient_source.js [hunk 10]
```diff
@@ -6,38 +6,41 @@ module.exports = class MongoDBPatientSource {
   /* Insert documentation here
     */
```
Comment (jbradl11) [inline lib/models/patient_source.js]: Should this be deleted?

File: lib/models/patient_source.js [hunk 11]
```diff
@@ -6,38 +6,41 @@ module.exports = class MongoDBPatientSource {
...
 
-  constructor(connectionInfo) {
-    mongoose.connect(connectionInfo);
+  constructor(connectionInfo = null) {
```
Comment (jbradl11) [inline lib/models/patient_source.js]: How will the "Patient Source", "Measure Source", and "ValueSet Source" work with running in the client? I am thinking there needs to be an optional parameter so that the Bonnie team can pass in data we want to use during execution. Thoughts?

File: lib/models/value_set_source.js [hunk 12]
```diff
@@ -0,0 +1,72 @@
+const { ValueSetSchema } = require('cqm-models');
+const mongoose = require('mongoose');
+const _ = require('lodash');
+
+module.exports = class MongoDBValueSetSource {
+  /* Insert documentation here
+    */
+
+  constructor(connectionInfo) {
```
Comment (jbradl11) [inline lib/models/value_set_source.js]: The Measure Source constructor has `connectionInfo = null`, should this constructor act like the Measure source constructor?

File: lib/models/value_set_source.js [hunk 13]
```diff
@@ -0,0 +1,72 @@
+const { ValueSetSchema } = require('cqm-models');
+const mongoose = require('mongoose');
+const _ = require('lodash');
+
+module.exports = class MongoDBValueSetSource {
+  /* Insert documentation here
```
Comment (jbradl11) [inline lib/models/value_set_source.js]: Should this be removed?
